rust/lance/src/io/writer/statistics.rs

```diff
@@ -0,0 +1,1082 @@
+        let split_point = idx + c.len_utf8();
+        if split_point <= length {
+            return data.as_bytes()[0..split_point].to_vec().into();
+        }
+    }
+
+    None
+}
+
+/// Truncate a binary slice to make sure its length is less than `length`
+fn truncate_binary(data: &[u8], length: usize) -> Option<Vec<u8>> {
+    // We return values like that at an earlier stage in the process.
+    assert!(data.len() >= length);
+    // If all bytes are already maximal, no need to truncate
+
+    data[0..length].to_vec().into()
+}
+
+/// Try and increment the bytes from right to left.
+///
+/// Returns `None` if all bytes are set to `u8::MAX`.
+fn increment(mut data: Vec<u8>) -> Option<Vec<u8>> {
+    for byte in data.iter_mut().rev() {
+        let (incremented, overflow) = byte.overflowing_add(1);
+        *byte = incremented;
+
+        if !overflow {
+            return Some(data);
+        }
+    }
+
+    None
+}
+
+/// Try and increment the the string's bytes from right to left, returning when the result is a valid UTF8 string.
+/// Returns `None` when it can't increment any byte.
+fn increment_utf8(mut data: Vec<u8>) -> Option<Vec<u8>> {
+    for idx in (0..data.len()).rev() {
+        let original = data[idx];
+        let (mut byte, mut overflow) = data[idx].overflowing_add(1);
+
+        // Until overflow: 0xFF -> 0x00
+        while !overflow {
+            data[idx] = byte;
+
+            if str::from_utf8(&data).is_ok() {
+                return Some(data);
+            }
+            (byte, overflow) = data[idx].overflowing_add(1);
+        }
+
+        data[idx] = original;
+    }
+
+    None
+}
+
+fn truncate_min_value(data: &[u8]) -> Vec<u8> {
+    match str::from_utf8(data) {
+        Ok(str_data) => truncate_utf8(str_data, BINARY_PREFIX_LENGTH),
+        Err(_) => truncate_binary(data, BINARY_PREFIX_LENGTH),
+    }
+    .unwrap_or_else(|| data.to_vec())
+}
+
+fn truncate_max_value(data: &[u8]) -> Vec<u8> {
+    match str::from_utf8(data) {
+        Ok(str_data) => truncate_utf8(str_data, BINARY_PREFIX_LENGTH).and_then(increment_utf8),
+        Err(_) => truncate_binary(data, BINARY_PREFIX_LENGTH).and_then(increment),
+    }
+    .unwrap_or_else(|| data.to_vec())
+}
+
+fn get_string_statistics(arrays: &[&ArrayRef]) -> StatisticsRow {
+    let mut min_value = vec![u8::MAX; BINARY_PREFIX_LENGTH];
+    let mut max_value = vec![u8::MIN; BINARY_PREFIX_LENGTH];
+    let mut null_count: i64 = 0;
+    let mut total_count: i64 = 0;
+
+    let arr = arrays
+        .iter()
+        .map(|x| x.as_any().downcast_ref::<StringArray>().unwrap())
+        .collect::<Vec<_>>();
+
+    for array in arr.iter() {
+        total_count += array.len() as i64;
+        array.iter().for_each(|value| {
+            if let Some(value) = value {
+                let value = value.as_bytes();
+                if let Some(Ordering::Greater) = value.partial_cmp(max_value.as_slice()) {
+                    max_value = value.to_vec();
+                }
```

Comment:
This seems to be comparing bytes and not strings, which seems wrong.

I think we may also save time by only comparing the first 16 bytes and not making a data copy every time we find a new value. 

```suggestion
                let value = truncate_utf8(value);
                if let Some(Ordering::Greater) = value.partial_cmp(max_value)) {
                    max_value = value;
                }
```